File: examples/deprecated/msg/network-ns3/network-ns3.c
/* Copyright (c) 2007-2020. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include "simgrid/msg.h"

#include <stdio.h> /* snprintf */

XBT_LOG_NEW_DEFAULT_CATEGORY(msg_test, "Messages specific for this msg example");

int timer_start; //set as 1 in the master process

//keep a pointer to all surf running tasks.
#define NTASKS 1500
double start_time, end_time, elapsed_time;
double gl_data_size[NTASKS];
msg_task_t gl_task_array[NTASKS];
const char *workernames[NTASKS];
const char *masternames[NTASKS];
int gl_task_array_id = 0;
int count_finished = 0;

#define FINALIZE ((void*)221297)        /* a magic number to tell people to stop working */

static int master(int argc, char *argv[])
{
  msg_task_t todo;

  xbt_assert(argc==4,"Strange number of arguments expected 3 got %d", argc - 1);

  XBT_DEBUG ("Master started");

  /* data size */
  double task_comm_size = xbt_str_parse_double(argv[1], "Invalid task communication size: %s");

  /* worker name */
  char *workername = argv[2];
  int id = xbt_str_parse_int(argv[3], "Invalid ID as argument 3: %s");   //unique id to control statistics
  char *id_alias = xbt_malloc(20*sizeof(char));
  snprintf(id_alias, 20, "flow_%d", id);
  workernames[id] = workername;
  TRACE_category(id_alias);

  masternames[id] = MSG_host_get_name(MSG_host_self());

  {                             /*  Task creation.  */
    todo = MSG_task_create("Task_0", 100*task_comm_size, task_comm_size, NULL);
    MSG_task_set_category(todo, id_alias);
    //keep track of running tasks
    gl_task_array[id] = todo;
    gl_data_size[id] = task_comm_size;
  }

  MSG_host_by_name(workername);

  count_finished++;
  timer_start = 1 ;

  /* time measurement */
  snprintf(id_alias,20,"%d", id);
  start_time = MSG_get_clock();
  MSG_task_send(todo, id_alias);
  end_time = MSG_get_clock();

  XBT_DEBUG ("Finished");
  xbt_free(id_alias);
  return 0;
}

static int timer(int argc, char *argv[])
{
  double sleep_time;
  double first_sleep;

  xbt_assert(argc==3,"Strange number of arguments expected 2 got %d", argc - 1);

  sscanf(argv[1], "%lf", &first_sleep);
  sscanf(argv[2], "%lf", &sleep_time);

  XBT_DEBUG ("Timer started");

  if(first_sleep){
      MSG_process_sleep(first_sleep);
  }

  do {
    XBT_DEBUG ("Get sleep");
    MSG_process_sleep(sleep_time);
  } while(timer_start);

  XBT_DEBUG ("Finished");
  return 0;
}

static int worker(int argc, char *argv[])
{
  msg_task_t task = NULL;
  char id_alias[10];

  xbt_assert(argc==2,"Strange number of arguments expected 1 got %d", argc - 1);

  XBT_DEBUG ("Worker started");

  int id = xbt_str_parse_int(argv[1], "Invalid id: %s");
  snprintf(id_alias,10, "%d", id);

  msg_error_t a = MSG_task_receive(&(task), id_alias);

  count_finished--;
  if(count_finished == 0){
      timer_start = 0;
  }

  xbt_assert(a == MSG_OK,"Hey?! What's up? Unexpected behavior");

  elapsed_time = MSG_get_clock() - start_time;

  XBT_INFO("FLOW[%d] : Receive %.0f bytes from %s to %s", id, MSG_task_get_bytes_amount(task), masternames[id],
           workernames[id]);
//  MSG_task_execute(task);
  MSG_task_destroy(task);

  XBT_DEBUG ("Finished");
  return 0;
}

int main(int argc, char *argv[])
{
  MSG_init(&argc, argv);
  xbt_assert(argc > 2, "Usage: %s platform_file deployment_file\n"
             "\tExample: %s platform.xml deployment.xml\n", argv[0], argv[0]);

  MSG_create_environment(argv[1]);
  TRACE_declare_mark("endmark");

  MSG_function_register("master", master);
  MSG_function_register("worker", worker);
  MSG_function_register("timer", timer);

  MSG_launch_application(argv[2]);

  msg_error_t res = MSG_main();

  return res != MSG_OK;
}

File: examples/java/network_ns3/network_ns3.java
/* Copyright (c) 2007-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

import java.util.HashMap;
import java.util.Map;
import org.simgrid.s4u.*;

class MasterWorkerNames {
  String master;
  String worker;
  MasterWorkerNames(String master, String worker)
  {
    this.master = master;
    this.worker = worker;
  }
}

class Payload {
  double msg_size;
  double start_time;
  Payload(double msg_size, double start_time)
  {
    this.msg_size   = msg_size;
    this.start_time = start_time;
  }
}

class Master extends Actor {
  static Map<Integer, MasterWorkerNames> names = new HashMap<>();

  String[] args;
  public Master(String[] args) { this.args = args; }

  public void run() throws SimgridException
  {
    if (args.length != 3)
      Engine.die("Strange number of arguments, expected 3 got %d", args.length);

    Engine.debug("Master started");

    /* data size */
    double msg_size = Double.parseDouble(args[0]);
    int id          = Integer.parseInt(args[2]); // unique id to control statistics

    /* master and worker names */
    names.putIfAbsent(id, new MasterWorkerNames(this.get_host().get_name(), args[1]));

    Mailbox mbox = this.get_engine().mailbox_by_name(args[2]);

    Payload payload = new Payload(msg_size, Engine.get_clock());
    mbox.put(payload, (long)msg_size);

    Engine.debug("Finished");
  }
}

class Worker extends Actor {
  String[] args;
  public Worker(String[] args) { this.args = args; }

  public void run() throws SimgridException
  {
    if (args.length != 1)
      Engine.die("Strange number of arguments, expected 1 got %d", args.length);

    int id       = Integer.parseInt(args[0]);
    Mailbox mbox = this.get_engine().mailbox_by_name(args[0]);

    Engine.debug("Worker started");

    Payload payload = (Payload)mbox.get();

    double elapsed_time = Engine.get_clock() - payload.start_time;

    MasterWorkerNames mw = Master.names.get(id);
    Engine.info("FLOW[%d] : Receive %.0f bytes from %s to %s", id, payload.msg_size, mw.master, mw.worker);
    Engine.debug("FLOW[%d] : transferred in %f seconds", id, elapsed_time);

    Engine.debug("Finished");
  }
}

public class network_ns3 {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);
    if (args.length < 2)
      Engine.die("Usage: network_ns3 platform_file deployment_file\n"
                 + "\tExample: platform.xml deployment.xml\n");

    e.load_platform(args[0]);
    e.load_deployment(args[1]);

    e.run();

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.delete(); // We need a synchronous, same-thread native teardown for ns3
  }
}
